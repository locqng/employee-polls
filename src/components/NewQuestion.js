import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { handleCreateQuestion } from "../actions/questions";
import { useNavigate, useLocation } from "react-router-dom";

const NewQuestions = (props) => {
  const navigate = useNavigate();
  const location = useLocation();
  const currentPath = location.pathname;

  const { dispatch, authedUser } = props;
  const [optOneText, setOptOneText] = useState("");
  const [optTwoText, setOptTwoText] = useState("");
  const [disableButton, setDisableButton] = useState(false);

  useEffect(() => {
    if (authedUser === null) {
      navigate("/", { state: { previousPage: currentPath } });
    }
  }, []);

  const handleOptOneChange = (e) => {
    setOptOneText(e.target.value);
    setDisableButton(e.target.value.trim() === "");
  };

  const handleOptTwoChange = (e) => {
    setOptTwoText(e.target.value);
    setDisableButton(e.target.value.trim() === "");
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    const optionOneText = optOneText;
    const optionTwoText = optTwoText;
    setDisableButton(true);
    dispatch(
      handleCreateQuestion(
        { optionOneText, optionTwoText, author: authedUser },
        () => {
          setOptOneText("");
          setOptTwoText("");
          navigate("/home");
        }
      )
    );
  };

  return (
    <div className="center">
      <h1>Would you rather</h1>
      <form className="new-question-form" onSubmit={handleSubmit}>
        <h4>First Option</h4>
        <input
          value={optOneText}
          onChange={(e) => handleOptOneChange(e)}
          placeholder="Option One"
          data-testid="text-option-one"
        />
        <h4>Second Option</h4>
        <input
          value={optTwoText}
          onChange={(e) => handleOptTwoChange(e)}
          placeholder="Option Two"
          data-testid="text-option-two"
        />
        <button
          className="btn"
          type="submit"
          disabled={disableButton}
          data-testid="submit-button"
        >
          Submit
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ users, questions, authedUser }) => {
  return {
    questions,
    authedUser,
    users,
  };
};

export default connect(mapStateToProps)(NewQuestions);
