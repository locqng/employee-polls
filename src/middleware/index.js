import logger from "./logger";
import { thunk } from "redux-thunk";
import { Tuple } from "@reduxjs/toolkit";

export default () => new Tuple(thunk, logger);
