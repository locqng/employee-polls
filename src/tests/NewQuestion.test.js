import * as React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureMockStore from "redux-mock-store";
import * as router from "react-router";
import middleware from "../middleware";
import NewQuestion from "../components/NewQuestion";
import { thunk } from "redux-thunk";

describe("NewQuestion", () => {
  const mockStore = configureMockStore([thunk]);
  const initialState = {
    questions: {
      "8xf0y6ziyjabvozdd253nd": {
        id: "8xf0y6ziyjabvozdd253nd",
        author: "user1",
        timestamp: 1467166872634,
        optionOne: {
          votes: ["sarahedo"],
          text: "Build our new application with Javascript",
        },
        optionTwo: {
          votes: [],
          text: "Build our new application with Typescript",
        },
      },
    },
    authedUser: "user1",
    users: {
      user1: {
        id: "user1",
        password: "123",
        name: "User1",
        avatarURL: "https://tylermcginnis.com/would-you-rather/sarah.jpg",
        answers: {
          "8xf0y6ziyjabvozdd253nd": "optionOne",
        },
        questions: ["8xf0y6ziyjabvozdd253nd"],
      },
    },
  };
  const store = mockStore(initialState);

  it("matches the snapshot", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewQuestion />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it("adds successfully", async () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewQuestion />
        </BrowserRouter>
      </Provider>
    );
    const optionOne = component.getByTestId("text-option-one");
    const optionTwo = component.getByTestId("text-option-two");
    fireEvent.change(optionOne, { target: { value: "Eat before 8pm" } });
    fireEvent.change(optionTwo, { target: { value: "Eat after 8pm" } });
    const submitButton = component.getByTestId("submit-button");
    fireEvent.click(submitButton);
    expect(submitButton).toBeDisabled();
  });
});
