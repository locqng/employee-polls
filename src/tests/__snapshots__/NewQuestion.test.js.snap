// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewQuestion matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="center"
      >
        <h1>
          Would you rather
        </h1>
        <form
          class="new-question-form"
        >
          <h4>
            First Option
          </h4>
          <input
            data-testid="text-option-one"
            placeholder="Option One"
            value=""
          />
          <h4>
            Second Option
          </h4>
          <input
            data-testid="text-option-two"
            placeholder="Option Two"
            value=""
          />
          <button
            class="btn"
            data-testid="submit-button"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="center"
    >
      <h1>
        Would you rather
      </h1>
      <form
        class="new-question-form"
      >
        <h4>
          First Option
        </h4>
        <input
          data-testid="text-option-one"
          placeholder="Option One"
          value=""
        />
        <h4>
          Second Option
        </h4>
        <input
          data-testid="text-option-two"
          placeholder="Option Two"
          value=""
        />
        <button
          class="btn"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
